packages/flutter/lib/src/rendering/object.dart

```diff
@@ -6370,97 +6368,70 @@ final class _SemanticsGeometry {
+      childToCommonAncestor.add(childRenderObject);
     }
-
-    // Calculate transform.
     assert(childToCommonAncestor.length >= 2);
-    for (int i = childToCommonAncestor.length - 1; i > 0; i -= 1) {
-      childToCommonAncestor[i].applyPaintTransform(childToCommonAncestor[i - 1], transform);
-    }
+    assert(identical(childRenderObject, parentRenderObject));
 
-    if (parentToCommonAncestorTransform != null) {
-      if (parentToCommonAncestorTransform.invert() != 0) {
-        transform.multiply(parentToCommonAncestorTransform);
-      } else {
-        transform.setZero();
-      }
-    }
+    // Calculate clips and transform.
 
-    // Calculate clips.
     Rect? paintClipRect;
     Rect? semanticsClipRect;
-    if (childToCommonAncestor.last == parent.renderObject) {
-      // This is most common case, i.e. parent is the common ancestor.
-      paintClipRect = parentPaintClipRect;
-      semanticsClipRect = parentSemanticsClipRect;
-      assert(parentToCommonAncestorTransform == null);
-      for (int i = childToCommonAncestor.length - 1; i > 0; i -= 1) {
-        (paintClipRect, semanticsClipRect) = _computeClipRect(
-          childToCommonAncestor[i],
-          childToCommonAncestor[i - 1],
-          semanticsClipRect,
-          paintClipRect,
-        );
-      }
-    } else {
-      // Otherwise we have to find the closest ancestor RenderObject that
-      // has up-to-date semantics geometry and compute the clip rects from there.
-      //
-      // Currently it can only happen when the subtree contains an OverlayPortal.
-      final List<RenderObject> clipPath = <RenderObject>[child.renderObject];
+    final Matrix4 transform = Matrix4.identity();
+    // Traverse from `parent`'s render object to `child`'s.
+    for (int i = childToCommonAncestor.length - 1; i > 0; i -= 1) {
+      final RenderObject nodeParent = childToCommonAncestor[i];
+      final RenderObject node = childToCommonAncestor[i - 1];
 
-      RenderObject? ancestor = child.renderObject.parent;
-      while (ancestor != null && ancestor._semantics.cachedSemanticsNode == null) {
-        clipPath.add(ancestor);
-        ancestor = ancestor.parent;
-      }
-      final SemanticsNode? ancestorNode = ancestor?._semantics.cachedSemanticsNode;
-      paintClipRect = ancestorNode?.parentPaintClipRect;
-      semanticsClipRect = ancestorNode?.parentSemanticsClipRect;
-      if (ancestor != null) {
-        RenderObject parent = ancestor;
-        for (int i = clipPath.length - 1; i >= 0; i -= 1) {
-          (paintClipRect, semanticsClipRect) = _computeClipRect(
-            parent,
-            clipPath[i],
-            semanticsClipRect,
-            paintClipRect,
-          );
-          parent = clipPath[i];
-        }
-      }
+      final Rect? localPaintClipInParent = _transformRect(
+        nodeParent.describeApproximatePaintClip(node),
+        transform, // paint transform from nodeParent to parent
```

Comment:
the comment seems to be inaccurate. Isn't this transform going to be from original child to the current ancestor?